flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/table/KeyedStateReader.java

```diff
@@ -0,0 +1,362 @@
+
+            case DECIMAL: // BigDecimal, ByteBuffer, byte[]
+                return convertToDecimal(object, logicalType);
+
+            case TINYINT: // Byte
+            case SMALLINT: // Short
+            case INTEGER: // Integer
+            case BIGINT: // Long
+            case FLOAT: // Float
+            case DOUBLE: // Double
+            case DATE: // Integer
+                return object;
+
+            case INTERVAL_YEAR_MONTH: // Long
+            case INTERVAL_DAY_TIME: // Long
+                return object;
+
+            case ARRAY:
+                return convertToArray(object, logicalType);
+
+            case MAP:
+                return convertToMap(object, logicalType);
+
+            case ROW:
+                return convertToRow(object, logicalType);
+
+            case NULL:
+                return null;
+
+            case MULTISET:
+            case TIME_WITHOUT_TIME_ZONE:
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+            case TIMESTAMP_WITH_TIME_ZONE:
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+            case DISTINCT_TYPE:
+            case STRUCTURED_TYPE:
+            case RAW:
+            case SYMBOL:
+            case UNRESOLVED:
+            default:
+                throw new UnsupportedOperationException("Unsupported type: " + logicalType);
+        }
+    }
+
+    private Object getObjectField(Object object, RowType.RowField rowField) {
+        String rowFieldName = rowField.getName();
+
+        Class objectClass = object.getClass();
+        Object objectField;
+        try {
+            Field field = objectClass.getField(rowFieldName);
+            objectField = field.get(object);
+        } catch (NoSuchFieldException e1) {
+            Method method = getMethod(rowFieldName, objectClass);
+            try {
+                objectField = method.invoke(object);
+            } catch (IllegalAccessException | InvocationTargetException e2) {
+                throw new RuntimeException(e2);
+            }
+        } catch (IllegalAccessException e) {
+            throw new UnsupportedOperationException(
+                    "Cannot access field by either public member or getter function: "
+                            + rowField.getName());
+        }
+
+        return objectField;
+    }
+
+    private Method getMethod(String rowFieldName, Class objectClass) {
+        String upperRowFieldName =
+                rowFieldName.substring(0, 1).toUpperCase() + rowFieldName.substring(1);
+        Method method;
+        try {
+            String methodName = "get" + upperRowFieldName;
+            method = objectClass.getMethod(methodName);
+        } catch (NoSuchMethodException e2) {
+            try {
+                String methodName = "is" + upperRowFieldName;
+                method = objectClass.getMethod(methodName);
+            } catch (NoSuchMethodException e3) {
+                throw new RuntimeException(e3);
+            }
+        }
+        return method;
+    }
```

Comment:
IIUC this is using the java reflection to access the sub field of a pojo, right? And it is invoked each time an incoming instance should be converted, which is slow. Is it possible to maintain a class/method cache, or a converter, to avoid calling reflection on the fly?

Reply:
Just added guava cache to have a limited entries solution.